File: scripts/ci/support/AtlasVaultTwoDeviceConvergenceClient.swift
import CryptoKit
import Foundation

private let collectionID = "collection_c20"

private func key(_ label: String) -> Data {
    Data(SHA256.hash(data: Data("atlasvault-c20-synthetic-\(label)".utf8)))
}

private func load(_ path: String) throws -> [String: Any] {
    guard let value = try JSONSerialization.jsonObject(
        with: Data(contentsOf: URL(fileURLWithPath: path))
    ) as? [String: Any] else {
        throw ClientError.invalid
    }
    return value
}

private func write(_ path: String, _ value: [String: Any]) throws {
    var data = try JSONSerialization.data(withJSONObject: value, options: [.sortedKeys])
    data.append(0x0a)
    try data.write(to: URL(fileURLWithPath: path), options: .atomic)
}

private enum ClientError: Error {
    case invalid
}

private struct Parts {
    let replica: AtlasVaultDurableEncryptedConvergentReplica
    let outbox: AtlasVaultDurableEncryptedOutbox
    let inbox: AtlasVaultDurableEncryptedInbox
}

private func parts(_ directory: URL) throws -> Parts {
    try FileManager.default.createDirectory(
        at: directory,
        withIntermediateDirectories: true,
        attributes: [.posixPermissions: 0o700]
    )
    return try Parts(
        replica: AtlasVaultDurableEncryptedConvergentReplica(
            fileURL: directory.appendingPathComponent("replica.state"),
            encryptionKey: key("replica"),
            authenticationKey: key("snapshot-authentication"),
            collectionID: collectionID
        ),
        outbox: AtlasVaultDurableEncryptedOutbox(
            fileURL: directory.appendingPathComponent("outbox.state"),
            encryptionKey: key("outbox")
        ),
        inbox: AtlasVaultDurableEncryptedInbox(
            fileURL: directory.appendingPathComponent("inbox.state"),
            encryptionKey: key("inbox")
        )
    )
}

private func operations(
    _ path: String
) throws -> [String: AtlasVaultEncryptedPatchOperation] {
    guard let raw = try load(path)["operations"] as? [String: [String: Any]] else {
        throw ClientError.invalid
    }
    return try raw.mapValues(AtlasVaultEncryptedPatchOperation.init(jsonObject:))
}

private func decryptTransport(
    _ operation: AtlasVaultEncryptedPatchOperation
) throws -> AtlasVaultEncryptedPatchOperation {
    guard let nonce = Data(base64Encoded: operation.envelope.nonceBase64),
          let combined = Data(base64Encoded: operation.envelope.ciphertextBase64),
          let aad = Data(base64Encoded: operation.envelope.aadBase64),
          combined.count > 16
    else {
        throw ClientError.invalid
    }
    let sealed = try AES.GCM.SealedBox(
        nonce: AES.GCM.Nonce(data: nonce),
        ciphertext: combined.dropLast(16),
        tag: combined.suffix(16)
    )
    let clear = try AES.GCM.open(sealed, using: SymmetricKey(data: key("transport")), authenticating: aad)
    guard let json = try JSONSerialization.jsonObject(with: clear) as? [String: Any] else {
        throw ClientError.invalid
    }
    return try AtlasVaultEncryptedPatchOperation(jsonObject: json)
}

private func result(_ parts: Parts) throws -> [String: Any] {
    let records = try parts.replica.currentRecords().map(\.jsonObject)
    let canonical = try JSONSerialization.data(
        withJSONObject: records,
        options: [.sortedKeys, .withoutEscapingSlashes]
    )
    return [
        "pid": ProcessInfo.processInfo.processIdentifier,
        "records": records,
        "state_sha256": SHA256.hash(data: canonical)
            .map { String(format: "%02x", $0) }.joined(),
        "accepted_operation_count": try parts.replica.acceptedOperationCount(),
        "pending_replica_count": try parts.replica.pendingOperations().count,
        "pending_outbox_count": try parts.outbox.pendingOperations().count,
        "cursor": try parts.inbox.cursor() ?? NSNull(),
    ]
}

private func prepare(_ arguments: [String]) throws {
    guard arguments.count == 5 else { throw ClientError.invalid }
    let all = try operations(arguments[2])
    let parts = try parts(URL(fileURLWithPath: arguments[1], isDirectory: true))
    for name in arguments[3].split(separator: ",").map(String.init) {
        guard let operation = all[name] else { throw ClientError.invalid }
        _ = try parts.replica.queueLocal(operation)
        try parts.outbox.enqueue(operation)
    }
    var output = try result(parts)
    output["operations"] = try parts.outbox.pendingOperations().map(\.jsonObject)
    try write(arguments[4], output)
}

private func apply(_ arguments: [String]) throws {
    guard arguments.count == 4 else { throw ClientError.invalid }
    let page = try load(arguments[2])
    let parts = try parts(URL(fileURLWithPath: arguments[1], isDirectory: true))
    let pendingReplica = Set(try parts.replica.pendingOperations().map(\.operationID))
    let pendingOutbox = Set(try parts.outbox.pendingOperations().map(\.operationID))
    guard let accepted = page["accepted_operation_ids"] as? [String],
          let rawOperations = page["transport_operations"] as? [[String: Any]]
    else {
        throw ClientError.invalid
    }
    for operationID in accepted {
        if pendingReplica.contains(operationID) {
            try parts.replica.confirmRemoteAcceptance(operationID)
        }
        if pendingOutbox.contains(operationID) {
            try parts.outbox.confirmRemoteAcceptance(operationID)
        }
    }
    try parts.inbox.stagePage(
        expectedCursor: page["expected_cursor"] is NSNull ? nil : page["expected_cursor"] as? String,
        nextCursor: page["next_cursor"] is NSNull ? nil : page["next_cursor"] as? String,
        operations: try rawOperations.map(
            AtlasVaultEncryptedPatchOperation.init(jsonObject:)
        )
    )
    while try parts.inbox.applyNext({ transport in
        _ = try parts.replica.ingestRemote(decryptTransport(transport))
    }) != nil {}
    try write(arguments[3], result(parts))
}

@main
private struct AtlasVaultTwoDeviceConvergenceClient {
    static func main() {
        do {
            let arguments = Array(CommandLine.arguments.dropFirst())
            guard let action = arguments.first else { throw ClientError.invalid }
            if action == "prepare" {
                try prepare(arguments)
            } else if action == "apply" {
                try apply(arguments)
            } else {
                throw ClientError.invalid
            }
        } catch {
            FileHandle.standardError.write(Data("C20 client failed: \(error)\n".utf8))
            exit(1)
        }
    }
}

File: apps/atlas_flutter/test/support/atlas_vault_two_device_convergence_client.dart
import 'dart:collection';
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:atlas/src/atlas_vault/sync_queue.dart';
import 'package:cryptography/cryptography.dart';

const _collectionId = 'collection_c20';

Future<Uint8List> _key(String label) async => Uint8List.fromList(
  (await Sha256().hash(utf8.encode('atlasvault-c20-synthetic-$label'))).bytes,
);

Future<Map<String, Object?>> _load(String path) async =>
    (jsonDecode(await File(path).readAsString()) as Map)
        .cast<String, Object?>();

Future<void> _write(String path, Map<String, Object?> value) async {
  await File(
    path,
  ).writeAsString('${jsonEncode(_canonical(value))}\n', flush: true);
}

Object? _canonical(Object? value) {
  if (value is Map) {
    final ordered = SplayTreeMap<String, Object?>();
    for (final entry in value.entries) {
      ordered[entry.key as String] = _canonical(entry.value);
    }
    return ordered;
  }
  if (value is List) return value.map(_canonical).toList(growable: false);
  return value;
}

Future<
  (
    AtlasVaultDurableEncryptedConvergentReplica,
    AtlasVaultDurableEncryptedOutbox,
    AtlasVaultDurableEncryptedInbox,
  )
>
_parts(Directory directory) async {
  await directory.create(recursive: true);
  final replica = AtlasVaultDurableEncryptedConvergentReplica(
    File('${directory.path}/replica.state'),
    encryptionKey: await _key('replica'),
    authenticationKey: await _key('snapshot-authentication'),
    collectionId: _collectionId,
  );
  final outbox = AtlasVaultDurableEncryptedOutbox(
    File('${directory.path}/outbox.state'),
    encryptionKey: await _key('outbox'),
  );
  final inbox = AtlasVaultDurableEncryptedInbox(
    File('${directory.path}/inbox.state'),
    encryptionKey: await _key('inbox'),
  );
  return (replica, outbox, inbox);
}

Future<Map<String, AtlasVaultEncryptedPatchOperation>> _operations(
  String path,
) async {
  final root = await _load(path);
  final raw = (root['operations']! as Map).cast<String, Object?>();
  return <String, AtlasVaultEncryptedPatchOperation>{
    for (final entry in raw.entries)
      entry.key: AtlasVaultEncryptedPatchOperation.fromJson(
        (entry.value! as Map).cast<String, Object?>(),
      ),
  };
}

Future<AtlasVaultEncryptedPatchOperation> _decryptTransport(
  AtlasVaultEncryptedPatchOperation operation,
) async {
  final encrypted = base64Decode(operation.envelope.ciphertextBase64);
  final clear = await AesGcm.with256bits().decrypt(
    SecretBox(
      encrypted.sublist(0, encrypted.length - 16),
      nonce: base64Decode(operation.envelope.nonceBase64),
      mac: Mac(encrypted.sublist(encrypted.length - 16)),
    ),
    secretKey: SecretKey(await _key('transport')),
    aad: base64Decode(operation.envelope.aadBase64),
  );
  return AtlasVaultEncryptedPatchOperation.fromJson(
    (jsonDecode(utf8.decode(clear)) as Map).cast<String, Object?>(),
  );
}

Future<Map<String, Object?>> _result(
  AtlasVaultDurableEncryptedConvergentReplica replica,
  AtlasVaultDurableEncryptedOutbox outbox,
  AtlasVaultDurableEncryptedInbox inbox,
) async {
  final records = (await replica.currentRecords())
      .map((item) => item.toJson())
      .toList(growable: false);
  final encoded = utf8.encode(jsonEncode(_canonical(records)));
  final digest = await Sha256().hash(encoded);
  return <String, Object?>{
    'pid': pid,
    'records': records,
    'state_sha256': digest.bytes
        .map((value) => value.toRadixString(16).padLeft(2, '0'))
        .join(),
    'accepted_operation_count': await replica.acceptedOperationCount(),
    'pending_replica_count': (await replica.pendingOperations()).length,
    'pending_outbox_count': (await outbox.pendingOperations()).length,
    'cursor': await inbox.readCursor(),
  };
}

Future<void> _prepare(List<String> arguments) async {
  if (arguments.length != 5) exit(64);
  final operations = await _operations(arguments[2]);
  final (replica, outbox, inbox) = await _parts(Directory(arguments[1]));
  for (final name in arguments[3].split(',')) {
    await replica.queueLocal(operations[name]!);
    await outbox.enqueue(operations[name]!);
  }
  final result = await _result(replica, outbox, inbox);
  result['operations'] = (await outbox.pendingOperations())
      .map((item) => item.toJson())
      .toList(growable: false);
  await _write(arguments[4], result);
}

Future<void> _apply(List<String> arguments) async {
  if (arguments.length != 4) exit(64);
  final page = await _load(arguments[2]);
  final (replica, outbox, inbox) = await _parts(Directory(arguments[1]));
  final pendingReplica = (await replica.pendingOperations())
      .map((item) => item.operationId)
      .toSet();
  final pendingOutbox = (await outbox.pendingOperations())
      .map((item) => item.operationId)
      .toSet();
  for (final operationId
      in (page['accepted_operation_ids']! as List).cast<String>()) {
    if (pendingReplica.contains(operationId)) {
      await replica.confirmRemoteAcceptance(operationId);
    }
    if (pendingOutbox.contains(operationId)) {
      await outbox.confirmRemoteAcceptance(operationId);
    }
  }
  await inbox.stagePage(
    expectedCursor: page['expected_cursor'] as String?,
    nextCursor: page['next_cursor'] as String?,
    operations: (page['transport_operations']! as List).map(
      (item) => AtlasVaultEncryptedPatchOperation.fromJson(
        (item as Map).cast<String, Object?>(),
      ),
    ),
  );
  while (await inbox.applyNext((transport) async {
        await replica.ingestRemote(await _decryptTransport(transport));
      }) !=
      null) {}
  await _write(arguments[3], await _result(replica, outbox, inbox));
}

Future<void> main(List<String> arguments) async {
  if (arguments.isEmpty) exit(64);
  if (arguments.first == 'prepare') {
    await _prepare(arguments);
  } else if (arguments.first == 'apply') {
    await _apply(arguments);
  } else {
    exit(64);
  }
}
